pslab/bus/spi.py

```diff
@@ -0,0 +1,505 @@
+        if primary_prescaler not in range(0, 4):
+            error_message.append("Primary Prescaler must be in 2-bits.")
+        if secondary_prescaler not in range(0, 8):
+            error_message.append("Secondary Prescale must be in 3-bits.")
+        if CKE not in (0, 1):
+            error_message.append("Clock Edge Select must be a bit.")
+        if CKP not in (0, 1):
+            error_message.append("Clock Polarity must be a bit.")
+        if SMP not in (0, 1):
+            error_message.append("SMP must be a bit.")
+        if error_message:
+            raise ValueError("\n".join(error_message))
+
+        self._device.send_byte(CP.SPI_HEADER)
+        self._device.send_byte(CP.SET_SPI_PARAMETERS)
+        # 0Bhgfedcba - > <g>: modebit CKP,<f>: modebit CKE, <ed>:primary prescaler,
+        #                <cba>:secondary prescaler
+        self._device.send_byte(
+            secondary_prescaler
+            | (primary_prescaler << 3)
+            | (CKE << 5)
+            | (CKP << 6)
+            | (SMP << 7)
+        )
+        self._device.get_ack()
+        self._save_config(primary_prescaler, secondary_prescaler, CKE, CKP, SMP)
+
+    def _start(self):
+        """Select SPI channel to enable.
+
+        Basically sets the relevant chip select pin to LOW.
+
+        External ChipSelect pins:
+            version < 5 : {6, 7} # RC5, RC4 (dropped support)
+            version == 5 : {} (don't have any external CS pins)
+            version == 6 : {7} # RC4
+        """
+        self._device.send_byte(CP.SPI_HEADER)
+        self._device.send_byte(CP.START_SPI)
+        self._device.send_byte(7)  # SPI.CS v6
+        # No ACK because `RESPONSE == DO_NOT_BOTHER` in firmware.
+
+    def _stop(self):
+        """Select SPI channel to disable.
+
+        Sets the relevant chip select pin to HIGH.
+        """
+        self._device.send_byte(CP.SPI_HEADER)
+        self._device.send_byte(CP.STOP_SPI)
+        self._device.send_byte(7)  # SPI.CS v6
+
+    def _transfer(self, data: int, bits: int) -> int:
+        """Send data over SPI and receive data from SPI simultaneously.
+
+        Relevent SPI channel need to be enabled first.
+
+        It is a primitive SPI method, prefered to use :meth:`SPISlave.transfer8`
+        and :meth:`SPISlave.transfer16`.
+
+        Parameters
+        ----------
+        data : int
+            Data to transmit.
+        bits : int
+            The number of bits per word.
+
+        Returns
+        -------
+        data_in : int
+            Data returned by slave device.
+
+        Raises
+        ------
+        ValueError
+            If given bits per word not supported by PSLab board.
+        """
+        self._device.send_byte(CP.SPI_HEADER)
+        command = self._TRANSFER_COMMANDS_MAP.get(bits)
+
+        if not command:
+            raise ValueError(
+                f"PSLab only supports {set(self._TRANSFER_COMMANDS_MAP.keys())}"
+                + " bits per word."
+            )
```

Comment:
After sending a primary command byte, it is mandatory to send a secondary command byte. Sending only a single command byte puts the device in an undefined state. We should check that the `bits` parameter is valid before sending the primary command byte.